File: scraper.py
import re
import calendar
import feedparser
import requests
from requests.adapters import HTTPAdapter
from urllib3.util.retry import Retry
from bs4 import BeautifulSoup
from database import save_article, get_articles_by_source_urls
from config import (
    RSS_FEEDS, CATEGORY_KEYWORDS, CATEGORY_NEGATIVE_KEYWORDS, GOOGLE_NEWS_CATEGORIES,
    MAX_ARTICLE_AGE_DAYS, GN,
)
from datetime import datetime, timedelta, timezone
from concurrent.futures import ThreadPoolExecutor, as_completed
from urllib.parse import urlparse, quote_plus
from googlenewsdecoder import new_decoderv1
import time

# Shared session with automatic retry/backoff for transient network errors
# (connection resets, timeouts, and 429/5xx) so a single flaky host doesn't
# fail scraping outright.
_session = requests.Session()
_retry = Retry(
    total=3,
    backoff_factor=0.5,
    status_forcelist=[429, 500, 502, 503, 504],
    allowed_methods=["GET"],
)
_session.mount("https://", HTTPAdapter(max_retries=_retry))
_session.mount("http://", HTTPAdapter(max_retries=_retry))

PAYWALLED_DOMAINS = {
    "nytimes.com", "wsj.com", "washingtonpost.com", "ft.com",
    "theatlantic.com", "thetimes.co.uk", "telegraph.co.uk",
    "economist.com", "newyorker.com", "bloomberg.com",
    "hbr.org", "businessinsider.com", "insider.com",
    "latimes.com", "bostonglobe.com", "sfchronicle.com",
}

HEADERS = {"User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 Chrome/120 Safari/537.36"}


def resolve_url(url: str) -> str:
    """Decode Google News URL to the real article URL."""
    if "news.google.com" not in url:
        return url
    try:
        result = new_decoderv1(url, interval=2)
        if result.get("status") and result.get("decoded_url"):
            return result["decoded_url"]
    except Exception:
        pass
    return url


def domain_to_name(url: str) -> str:
    """Convert a URL to a clean publisher name, e.g. 'startribune.com' → 'Star Tribune'."""
    try:
        domain = urlparse(url).netloc.lower().removeprefix("www.")
        # Strip common TLDs/subdomains for display
        name = domain.split(".")[0]
        return name.replace("-", " ").title()
    except Exception:
        return url


def fetch_og_image(url: str):
    """Fetch og:image from an article page.

    Some sites (e.g. WordPress-based ones like NBC New York) pack the <head>
    with tracking/analytics tags before the og:image meta tag, pushing it well
    past a small byte budget — read until </head> (with a generous safety
    cap) instead of cutting off early, or the meta tag gets missed entirely.
    """
    try:
        resp = _session.get(url, timeout=8, headers=HEADERS, stream=True)
        content = b""
        for chunk in resp.iter_content(8192):
            content += chunk
            if b"</head>" in content or b"og:image" in content or len(content) > 500_000:
                break
        soup = BeautifulSoup(content, "html.parser")
        tag = soup.find("meta", property="og:image") or soup.find("meta", attrs={"name": "og:image"})
        if tag and tag.get("content"):
            return tag["content"]
    except Exception:
        pass
    return None


def is_paywalled(url: str) -> bool:
    domain = urlparse(url).netloc.lower().removeprefix("www.")
    return any(domain == p or domain.endswith("." + p) for p in PAYWALLED_DOMAINS)


def fetch_article_body(url: str) -> str:
    """Scrape full article body text when RSS provides no excerpt."""
    try:
        resp = _session.get(url, timeout=6, headers=HEADERS, stream=True)
        content = b""
        for chunk in resp.iter_content(8192):
            content += chunk
            if len(content) > 200000:
                break
        soup = BeautifulSoup(content, "html.parser")

        # Remove noise tags
        for tag in soup(["script", "style", "nav", "header", "footer", "aside", "form", "ads"]):
            tag.decompose()

        # Try common article content containers in order of specificity
        body = (
            soup.find("article") or
            soup.find(attrs={"itemprop": "articleBody"}) or
            soup.find(class_=lambda c: c and any(x in c.lower() for x in ["article-body", "article__body", "story-body", "post-body", "entry-content", "article-content"])) or
            soup.find("main")
        )

        if body:
            paragraphs = body.find_all("p")
        else:
            paragraphs = soup.find_all("p")

        text = " ".join(p.get_text(strip=True) for p in paragraphs if len(p.get_text(strip=True)) > 40)
        return text[:1000] if text else ""
    except Exception:
        return ""


def enrich_article(article: dict, is_google_news: bool) -> dict:
    """Resolve real URL, check paywall, set real source name, fetch og:image and body."""
    url = article["url"]

    if is_google_news and "news.google.com" in url:
        resolved = resolve_url(url)
        if resolved and "news.google.com" not in resolved:
            article["url"] = resolved
            url = resolved
            # Set source name from actual domain if not already set by feedparser
            if not article.get("source") or article["source"] == "Google News":
                article["source"] = domain_to_name(url)

    if is_paywalled(url):
        article["_paywalled"] = True
        return article

    real_url = url and "news.google.com" not in url

    # Fetch full article body if RSS gave us no useful excerpt
    if real_url and len(article.get("excerpt") or "") < 100:
        body = fetch_article_body(url)
        if body:
            article["excerpt"] = body

    if not article.get("image_url") and real_url:
        article["image_url"] = fetch_og_image(url)

    return article


_TITLE_STOPWORDS = {
    "the", "a", "an", "in", "on", "at", "to", "of", "and", "for", "is", "are",
    "after", "with", "who", "his", "her", "their", "was", "has", "have",
}


def _title_words(title: str) -> set:
    return {
        w for w in re.findall(r"[a-z0-9]+", title.lower())
        if w not in _TITLE_STOPWORDS and len(w) > 2
    }


def _is_near_duplicate_title(title: str, seen_word_sets: list) -> bool:
    """Catch different publishers covering the same story with reworded
    headlines (word-overlap ratio) — not exhaustive, just a cheap first pass.
    """
    words = _title_words(title)
    if not words:
        return False
    for other in seen_word_sets:
        overlap = len(words & other) / len(words | other)
        if overlap >= 0.6:
            return True
    return False


def _parse_published(entry) -> datetime:
    """Return a timezone-aware UTC datetime from the entry's published date, or
    None if feedparser couldn't parse one."""
    parsed = entry.get("published_parsed")
    if not parsed:
        return None
    try:
        return datetime.fromtimestamp(calendar.timegm(parsed), tz=timezone.utc)
    except Exception:
        return None


def scrape_rss(category: str, urls: list) -> list:
    articles = []
    for feed_url in urls:
        try:
            # Fetch with an explicit timeout first — feedparser.parse(url) has no
            # timeout of its own and can hang indefinitely on a slow feed server.
            resp = _session.get(feed_url, timeout=8, headers=HEADERS)
            feed = feedparser.parse(resp.content)
            for entry in feed.entries[:10]:
                excerpt = ""
                if hasattr(entry, "summary"):
                    soup = BeautifulSoup(entry.summary, "html.parser")
                    excerpt = soup.get_text()[:500]

                published_dt = _parse_published(entry)
                # Google News search results aren't strictly date-sorted, so old
                # articles matching the query show up alongside fresh ones —
                # drop anything we can confirm is too old. Keep entries with no
                # parseable date rather than risk dropping legitimate current ones.
                if published_dt and published_dt < datetime.now(timezone.utc) - timedelta(days=MAX_ARTICLE_AGE_DAYS):
                    continue
                published = published_dt.isoformat() if published_dt else entry.get("published")

                image_url = None
                if hasattr(entry, "media_thumbnail") and entry.media_thumbnail:
                    image_url = entry.media_thumbnail[0].get("url")
                elif hasattr(entry, "media_content") and entry.media_content:
                    image_url = entry.media_content[0].get("url")
                elif hasattr(entry, "enclosures") and entry.enclosures:
                    for enc in entry.enclosures:
                        if enc.get("type", "").startswith("image"):
                            image_url = enc.get("url")
                            break

                # For Google News entries, prefer entry.source.value (real publisher name)
                source = ""
                if hasattr(entry, "source") and isinstance(entry.source, dict):
                    source = entry.source.get("value", "") or entry.source.get("title", "")
                if not source:
                    source = feed.feed.get("title", feed_url)

                # Strip " - Publisher Name" suffix Google News appends to titles
                title = entry.get("title", "")
                if source and title.endswith(f" - {source}"):
                    title = title[: -len(f" - {source}")]

                link = entry.get("link", "")
                articles.append({
                    "url": link,
                    "source_url": link,  # raw feed link, before Google News redirect resolution
                    "title": title,
                    "source": source,
                    "category": category,
                    "excerpt": excerpt,
                    "summary": None,
                    "published_at": published,
                    "image_url": image_url,
                })
            time.sleep(0.2)
        except Exception as e:
            print(f"  Failed to scrape {feed_url}: {e}")
    return articles


def _interest_matches_category(interest_lower: str, cat: str) -> bool:
    """Word-level match between a user interest and a feed category. The old
    character-substring test ("ai" in "ukraine") matched unrelated categories
    on shared letters — feeding a "ukraine" user AI articles while ALSO
    skipping the on-the-fly Google News feed for their actual topic (the
    fallback only fires when nothing matched)."""
    if interest_lower == cat:
        return True
    iw = set(re.findall(r"[a-z0-9]+", interest_lower))
    cw = set(re.findall(r"[a-z0-9]+", cat))
    return bool(iw) and bool(cw) and (iw <= cw or cw <= iw)


def scrape_all(interests: list) -> list:
    """Scrape feeds for a list of interest categories."""
    all_articles = []
    seen_urls = set()

    for interest in interests:
        interest_lower = interest.lower()
        matched = [
            (cat, urls) for cat, urls in RSS_FEEDS.items()
            if _interest_matches_category(interest_lower, cat)
        ]
        if not matched:
            # Not one of the predefined categories — build a Google News search
            # feed for it on the fly instead of silently falling back to
            # "world news" (which just re-shows articles the user already has
            # and drops their actual interest).
            matched = [(interest_lower, [GN + quote_plus(interest.strip())])]

        for cat, urls in matched:
            is_google = cat in GOOGLE_NEWS_CATEGORIES or any("news.google.com" in u for u in urls)
            print(f"  Scraping {cat} ({'Google News' if is_google else 'RSS'})...")
            articles = scrape_rss(cat, urls)

            negative_keywords = CATEGORY_NEGATIVE_KEYWORDS.get(cat, [])
            keywords = CATEGORY_KEYWORDS.get(cat, [])
            seen_sources_in_cat = {}
            seen_titles_in_cat = []

            candidates = []
            for a in articles:
                if not a["url"] or a["url"] in seen_urls:
                    continue
                text = (a["title"] + " " + a.get("excerpt", "")).lower()
                if not is_google and keywords:
                    if not any(kw.lower() in text for kw in keywords):
                        continue
                if negative_keywords:
                    if any(nkw.lower() in text for nkw in negative_keywords):
                        continue
                source = a.get("source", "")
                if seen_sources_in_cat.get(source, 0) >= 2:
                    continue
                if _is_near_duplicate_title(a["title"], seen_titles_in_cat):
                    continue
                seen_sources_in_cat[source] = seen_sources_in_cat.get(source, 0) + 1
                seen_titles_in_cat.append(_title_words(a["title"]))
                candidates.append(a)

            # Skip re-enriching (re-decoding Google News redirects, re-fetching
            # body/og:image) for candidates we've already scraped in a previous
            # run — reuse the cached row instead. This is what makes repeat
            # refreshes fast instead of re-doing full network work every time.
            known = get_articles_by_source_urls([c["source_url"] for c in candidates])
            to_enrich = []
            for c in candidates:
                cached = known.get(c["source_url"])
                if cached and cached.get("url") and cached["url"] not in seen_urls:
                    seen_urls.add(cached["url"])
                    all_articles.append({
                        "url": cached["url"],
                        "source_url": c["source_url"],
                        "title": cached["title"],
                        "source": cached["source"],
                        "category": cat,
                        "excerpt": cached["excerpt"],
                        "summary": cached["summary"],
                        "image_url": cached["image_url"],
                        "published_at": cached["published_at"],
                    })
                elif not cached:
                    to_enrich.append(c)

            print(f"  Enriching {len(to_enrich)} new articles ({len(candidates) - len(to_enrich)} already cached)...")
            with ThreadPoolExecutor(max_workers=3) as executor:
                futures = {executor.submit(enrich_article, a, is_google): a for a in to_enrich}
                for future in as_completed(futures):
                    try:
                        enriched = future.result()
                        if enriched.get("_paywalled"):
                            continue
                        if enriched["url"] and enriched["url"] not in seen_urls:
                            seen_urls.add(enriched["url"])
                            all_articles.append(enriched)
                    except Exception:
                        pass

    return all_articles


def save_articles(articles: list) -> list:
    ids = []
    for article in articles:
        article_id = save_article(article)
        if article_id:
            ids.append(article_id)
    print(f"  Saved {len(ids)} new articles.")
    return ids

File: scraper/scraper.go
// Package scraper fetches Google News / RSS feeds, decodes Google News
// redirect URLs to the real article, enriches entries with a body excerpt and
// og:image, and dedupes across sources. Ported from scraper.py.
package scraper

import (
	"context"
	"io"
	"net/http"
	"net/url"
	"regexp"
	"strings"
	"sync"
	"time"

	"github.com/PuerkitoBio/goquery"
	"github.com/mmcdole/gofeed/rss"

	"news-aggregator/config"
	"news-aggregator/db"
)

const userAgent = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 Chrome/120 Safari/537.36"

var paywalledDomains = map[string]bool{
	"nytimes.com": true, "wsj.com": true, "washingtonpost.com": true, "ft.com": true,
	"theatlantic.com": true, "thetimes.co.uk": true, "telegraph.co.uk": true,
	"economist.com": true, "newyorker.com": true, "bloomberg.com": true,
	"hbr.org": true, "businessinsider.com": true, "insider.com": true,
	"latimes.com": true, "bostonglobe.com": true, "sfchronicle.com": true,
}

// Scraper holds the shared HTTP client and DB handle used across a scrape run.
type Scraper struct {
	db     *db.DB
	client *http.Client
}

// New returns a Scraper backed by the given database.
func New(database *db.DB) *Scraper {
	return &Scraper{
		db: database,
		// No global timeout — each request sets its own via context, matching
		// the per-call timeouts in the Python version (8s feeds, 6s bodies).
		client: &http.Client{},
	}
}

// get performs a GET with the browser User-Agent, a per-request timeout, and
// retry/backoff on transient failures (connection errors and 429/5xx), so a
// single flaky host doesn't fail the whole scrape.
func (s *Scraper) get(rawURL string, timeout time.Duration) (*http.Response, error) {
	var lastErr error
	retryStatus := map[int]bool{429: true, 500: true, 502: true, 503: true, 504: true}
	for attempt := 0; attempt < 4; attempt++ {
		if attempt > 0 {
			// backoff_factor 0.5 → 0.5, 1.0, 2.0s
			time.Sleep(time.Duration(float64(time.Second) * 0.5 * float64(int(1)<<(attempt-1))))
		}
		ctx, cancel := context.WithTimeout(context.Background(), timeout)
		req, err := http.NewRequestWithContext(ctx, http.MethodGet, rawURL, nil)
		if err != nil {
			cancel()
			return nil, err
		}
		req.Header.Set("User-Agent", userAgent)
		resp, err := s.client.Do(req)
		if err != nil {
			cancel()
			lastErr = err
			continue
		}
		if retryStatus[resp.StatusCode] {
			resp.Body.Close()
			cancel()
			lastErr = &httpStatusError{resp.StatusCode}
			continue
		}
		// Caller reads and closes Body; cancel must outlive the read, so wrap
		// the body to cancel on Close.
		resp.Body = &cancelReadCloser{ReadCloser: resp.Body, cancel: cancel}
		return resp, nil
	}
	return nil, lastErr
}

type httpStatusError struct{ code int }

func (e *httpStatusError) Error() string { return http.StatusText(e.code) }

type cancelReadCloser struct {
	io.ReadCloser
	cancel context.CancelFunc
}

func (c *cancelReadCloser) Close() error {
	err := c.ReadCloser.Close()
	c.cancel()
	return err
}

// domainToName converts a URL to a clean publisher name, e.g.
// "startribune.com" → "Star Tribune".
func domainToName(rawURL string) string {
	u, err := url.Parse(rawURL)
	if err != nil || u.Host == "" {
		return rawURL
	}
	host := strings.TrimPrefix(strings.ToLower(u.Host), "www.")
	name := strings.SplitN(host, ".", 2)[0]
	name = strings.ReplaceAll(name, "-", " ")
	return titleCase(name)
}

// titleCase upper-cases the first letter of each word, like Python str.title().
func titleCase(s string) string {
	return strings.Join(mapWords(strings.Fields(s), func(w string) string {
		if w == "" {
			return w
		}
		return strings.ToUpper(w[:1]) + strings.ToLower(w[1:])
	}), " ")
}

func mapWords(in []string, f func(string) string) []string {
	out := make([]string, len(in))
	for i, w := range in {
		out[i] = f(w)
	}
	return out
}

func isPaywalled(rawURL string) bool {
	u, err := url.Parse(rawURL)
	if err != nil {
		return false
	}
	host := strings.TrimPrefix(strings.ToLower(u.Host), "www.")
	for p := range paywalledDomains {
		if host == p || strings.HasSuffix(host, "."+p) {
			return true
		}
	}
	return false
}

// fetchOGImage fetches the og:image from an article page. Some sites pack the
// <head> with tracking tags before og:image, so read until </head> (with a
// generous cap) rather than cutting off early.
func (s *Scraper) fetchOGImage(rawURL string) string {
	resp, err := s.get(rawURL, 8*time.Second)
	if err != nil {
		return ""
	}
	defer resp.Body.Close()
	content := readUntil(resp.Body, [][]byte{[]byte("</head>"), []byte("og:image")}, 500_000)
	doc, err := goquery.NewDocumentFromReader(strings.NewReader(string(content)))
	if err != nil {
		return ""
	}
	if img, ok := doc.Find(`meta[property="og:image"]`).Attr("content"); ok && img != "" {
		return img
	}
	if img, ok := doc.Find(`meta[name="og:image"]`).Attr("content"); ok && img != "" {
		return img
	}
	return ""
}

// readUntil reads from r until one of the stop markers appears in the buffer or
// maxBytes is reached, then returns the buffer.
func readUntil(r io.Reader, stops [][]byte, maxBytes int) []byte {
	var buf []byte
	chunk := make([]byte, 8192)
	for {
		n, err := r.Read(chunk)
		if n > 0 {
			buf = append(buf, chunk[:n]...)
			for _, stop := range stops {
				if containsBytes(buf, stop) {
					return buf
				}
			}
			if len(buf) > maxBytes {
				return buf
			}
		}
		if err != nil {
			return buf
		}
	}
}

func containsBytes(haystack, needle []byte) bool {
	return strings.Contains(string(haystack), string(needle))
}

var bodyContainerClasses = []string{
	"article-body", "article__body", "story-body", "post-body",
	"entry-content", "article-content",
}

// fetchArticleBody scrapes readable body text when RSS provides no excerpt.
func (s *Scraper) fetchArticleBody(rawURL string) string {
	resp, err := s.get(rawURL, 6*time.Second)
	if err != nil {
		return ""
	}
	defer resp.Body.Close()
	content := readUntil(resp.Body, nil, 200_000)
	doc, err := goquery.NewDocumentFromReader(strings.NewReader(string(content)))
	if err != nil {
		return ""
	}
	// Remove noise tags.
	doc.Find("script, style, nav, header, footer, aside, form, ads").Remove()

	// Pick the most specific content container available.
	var body *goquery.Selection
	if sel := doc.Find("article").First(); sel.Length() > 0 {
		body = sel
	} else if sel := doc.Find(`[itemprop="articleBody"]`).First(); sel.Length() > 0 {
		body = sel
	} else if sel := findByClassSubstr(doc, bodyContainerClasses); sel != nil && sel.Length() > 0 {
		body = sel
	} else if sel := doc.Find("main").First(); sel.Length() > 0 {
		body = sel
	}

	var paras *goquery.Selection
	if body != nil {
		paras = body.Find("p")
	} else {
		paras = doc.Find("p")
	}

	var parts []string
	paras.Each(func(_ int, p *goquery.Selection) {
		t := strings.TrimSpace(p.Text())
		if len(t) > 40 {
			parts = append(parts, t)
		}
	})
	text := strings.Join(parts, " ")
	if len(text) > 1000 {
		text = text[:1000]
	}
	return text
}

// findByClassSubstr returns the first element whose class attribute (lowercased)
// contains any of the target substrings.
func findByClassSubstr(doc *goquery.Document, substrs []string) *goquery.Selection {
	var found *goquery.Selection
	doc.Find("[class]").EachWithBreak(func(_ int, sel *goquery.Selection) bool {
		class, _ := sel.Attr("class")
		cl := strings.ToLower(class)
		for _, sub := range substrs {
			if strings.Contains(cl, sub) {
				found = sel
				return false
			}
		}
		return true
	})
	return found
}

// enrichArticle resolves the real URL, checks the paywall list, sets the real
// source name, and fetches og:image and body. Returns the (possibly mutated)
// article and whether it was found to be paywalled.
func (s *Scraper) enrichArticle(a *db.Article, isGoogleNews bool) (*db.Article, bool) {
	if isGoogleNews && strings.Contains(a.URL, "news.google.com") {
		resolved := s.resolveURL(a.URL)
		if resolved != "" && !strings.Contains(resolved, "news.google.com") {
			a.URL = resolved
			if a.Source == "" || a.Source == "Google News" {
				a.Source = domainToName(a.URL)
			}
		}
	}

	if isPaywalled(a.URL) {
		return a, true
	}

	realURL := a.URL != "" && !strings.Contains(a.URL, "news.google.com")

	if realURL && len(a.Excerpt) < 100 {
		if body := s.fetchArticleBody(a.URL); body != "" {
			a.Excerpt = body
		}
	}
	if a.ImageURL == "" && realURL {
		a.ImageURL = s.fetchOGImage(a.URL)
	}
	return a, false
}

var (
	wordRe        = regexp.MustCompile(`[a-z0-9]+`)
	titleStopword = map[string]bool{
		"the": true, "a": true, "an": true, "in": true, "on": true, "at": true,
		"to": true, "of": true, "and": true, "for": true, "is": true, "are": true,
		"after": true, "with": true, "who": true, "his": true, "her": true,
		"their": true, "was": true, "has": true, "have": true,
	}
)

// titleWords returns the significant words in a title (lowercased, no stopwords,
// length > 2) as a set.
func titleWords(title string) map[string]bool {
	out := map[string]bool{}
	for _, w := range wordRe.FindAllString(strings.ToLower(title), -1) {
		if !titleStopword[w] && len(w) > 2 {
			out[w] = true
		}
	}
	return out
}

// isNearDuplicateTitle catches different publishers covering the same story
// with reworded headlines via a Jaccard word-overlap ratio ≥ 0.6.
func isNearDuplicateTitle(title string, seen []map[string]bool) bool {
	words := titleWords(title)
	if len(words) == 0 {
		return false
	}
	for _, other := range seen {
		inter, union := 0, len(words)
		for w := range other {
			if words[w] {
				inter++
			} else {
				union++
			}
		}
		if union > 0 && float64(inter)/float64(union) >= 0.6 {
			return true
		}
	}
	return false
}

// interestMatchesCategory does a word-level match between a user interest and a
// feed category, avoiding the substring false-positives of the original test
// ("ai" in "ukraine").
func interestMatchesCategory(interestLower, cat string) bool {
	if interestLower == cat {
		return true
	}
	iw := wordSet(interestLower)
	cw := wordSet(cat)
	if len(iw) == 0 || len(cw) == 0 {
		return false
	}
	return isSubset(iw, cw) || isSubset(cw, iw)
}

func wordSet(s string) map[string]bool {
	out := map[string]bool{}
	for _, w := range wordRe.FindAllString(strings.ToLower(s), -1) {
		out[w] = true
	}
	return out
}

func isSubset(a, b map[string]bool) bool {
	for w := range a {
		if !b[w] {
			return false
		}
	}
	return true
}

// scrapeRSS fetches and parses one category's feed URLs into raw articles
// (pre-enrichment), applying the max-age filter.
func (s *Scraper) scrapeRSS(category string, urls []string) []*db.Article {
	var articles []*db.Article
	parser := &rss.Parser{}
	cutoff := time.Now().UTC().Add(-time.Duration(config.MaxArticleAgeDays) * 24 * time.Hour)

	for _, feedURL := range urls {
		resp, err := s.get(feedURL, 8*time.Second)
		if err != nil {
			continue
		}
		feed, perr := parser.Parse(resp.Body)
		resp.Body.Close()
		if perr != nil || feed == nil {
			continue
		}

		limit := len(feed.Items)
		if limit > 10 {
			limit = 10
		}
		for _, entry := range feed.Items[:limit] {
			excerpt := ""
			if entry.Description != "" {
				excerpt = htmlToText(entry.Description)
				if len(excerpt) > 500 {
					excerpt = excerpt[:500]
				}
			}

			published := ""
			if entry.PubDateParsed != nil {
				// Drop entries we can confirm are too old; keep undated ones.
				if entry.PubDateParsed.Before(cutoff) {
					continue
				}
				published = entry.PubDateParsed.UTC().Format(time.RFC3339)
			} else {
				published = entry.PubDate
			}

			imageURL := extractImage(entry)

			source := ""
			if entry.Source != nil {
				if entry.Source.Title != "" {
					source = entry.Source.Title
				} else {
					source = entry.Source.URL
				}
			}
			if source == "" {
				source = feed.Title
			}

			title := entry.Title
			if source != "" && strings.HasSuffix(title, " - "+source) {
				title = strings.TrimSuffix(title, " - "+source)
			}

			articles = append(articles, &db.Article{
				URL:         entry.Link,
				SourceURL:   entry.Link,
				Title:       title,
				Source:      source,
				Category:    category,
				Excerpt:     excerpt,
				PublishedAt: published,
				ImageURL:    imageURL,
				// Summary stays NULL until the summarizer runs.
			})
		}
		time.Sleep(200 * time.Millisecond)
	}
	return articles
}

// extractImage pulls an image URL from media:thumbnail, media:content, or an
// image enclosure, mirroring feedparser's preference order.
func extractImage(entry *rss.Item) string {
	if entry.Extensions != nil {
		if media, ok := entry.Extensions["media"]; ok {
			for _, key := range []string{"thumbnail", "content"} {
				if exts, ok := media[key]; ok {
					for _, e := range exts {
						if u := e.Attrs["url"]; u != "" {
							return u
						}
					}
				}
			}
		}
	}
	for _, enc := range entry.Enclosures {
		if strings.HasPrefix(enc.Type, "image") && enc.URL != "" {
			return enc.URL
		}
	}
	return ""
}

// htmlToText strips HTML markup and returns the text content.
func htmlToText(html string) string {
	doc, err := goquery.NewDocumentFromReader(strings.NewReader(html))
	if err != nil {
		return html
	}
	return doc.Text()
}

// ScrapeAll scrapes feeds for the given interest categories, deduping across
// sources and enriching new articles concurrently.
func (s *Scraper) ScrapeAll(interests []string) []*db.Article {
	var allArticles []*db.Article
	seenURLs := map[string]bool{}

	for _, interest := range interests {
		interestLower := strings.ToLower(interest)

		type catURLs struct {
			cat  string
			urls []string
		}
		var matched []catURLs
		for cat, urls := range config.RSSFeeds {
			if interestMatchesCategory(interestLower, cat) {
				matched = append(matched, catURLs{cat, urls})
			}
		}
		if len(matched) == 0 {
			// Not a predefined category — build a Google News search feed for it
			// rather than silently falling back to "world news".
			matched = []catURLs{{
				cat:  interestLower,
				urls: []string{config.GN + url.QueryEscape(strings.TrimSpace(interest))},
			}}
		}

		for _, m := range matched {
			cat, urls := m.cat, m.urls
			isGoogle := config.GoogleNewsCategories[cat] || anyContains(urls, "news.google.com")
			articles := s.scrapeRSS(cat, urls)

			negativeKeywords := config.CategoryNegativeKeywords[cat]
			keywords := config.CategoryKeywords[cat]
			seenSources := map[string]int{}
			var seenTitles []map[string]bool

			var candidates []*db.Article
			for _, a := range articles {
				if a.URL == "" || seenURLs[a.URL] {
					continue
				}
				text := strings.ToLower(a.Title + " " + a.Excerpt)
				if !isGoogle && len(keywords) > 0 {
					if !anyKeyword(text, keywords) {
						continue
					}
				}
				if len(negativeKeywords) > 0 && anyKeyword(text, negativeKeywords) {
					continue
				}
				if seenSources[a.Source] >= 2 {
					continue
				}
				if isNearDuplicateTitle(a.Title, seenTitles) {
					continue
				}
				seenSources[a.Source]++
				seenTitles = append(seenTitles, titleWords(a.Title))
				candidates = append(candidates, a)
			}

			// Reuse cached rows for candidates already scraped in a prior run,
			// so repeat refreshes skip re-decoding and re-fetching.
			sourceURLs := make([]string, len(candidates))
			for i, c := range candidates {
				sourceURLs[i] = c.SourceURL
			}
			known, _ := s.db.GetArticlesBySourceURLs(sourceURLs)

			var toEnrich []*db.Article
			for _, c := range candidates {
				cached, ok := known[c.SourceURL]
				if ok && cached.URL != "" && !seenURLs[cached.URL] {
					seenURLs[cached.URL] = true
					allArticles = append(allArticles, &db.Article{
						URL:         cached.URL,
						SourceURL:   c.SourceURL,
						Title:       cached.Title,
						Source:      cached.Source,
						Category:    cat,
						Excerpt:     cached.Excerpt,
						Summary:     cached.Summary,
						ImageURL:    cached.ImageURL,
						PublishedAt: cached.PublishedAt,
					})
				} else if !ok {
					toEnrich = append(toEnrich, c)
				}
			}

			// Enrich new articles concurrently (max 3 in flight), then collect
			// non-paywalled results under the dedup lock.
			enriched := s.enrichConcurrent(toEnrich, isGoogle)
			for _, e := range enriched {
				if e.URL != "" && !seenURLs[e.URL] {
					seenURLs[e.URL] = true
					allArticles = append(allArticles, e)
				}
			}
		}
	}
	return allArticles
}

// enrichConcurrent enriches articles with a bounded worker pool, dropping
// paywalled ones. Order is not preserved (matches the Python as_completed loop).
func (s *Scraper) enrichConcurrent(articles []*db.Article, isGoogle bool) []*db.Article {
	if len(articles) == 0 {
		return nil
	}
	sem := make(chan struct{}, 3)
	var wg sync.WaitGroup
	var mu sync.Mutex
	var out []*db.Article

	for _, a := range articles {
		wg.Add(1)
		go func(a *db.Article) {
			defer wg.Done()
			sem <- struct{}{}
			defer func() { <-sem }()
			enriched, paywalled := s.enrichArticle(a, isGoogle)
			if paywalled {
				return
			}
			mu.Lock()
			out = append(out, enriched)
			mu.Unlock()
		}(a)
	}
	wg.Wait()
	return out
}

func anyContains(list []string, sub string) bool {
	for _, s := range list {
		if strings.Contains(s, sub) {
			return true
		}
	}
	return false
}

func anyKeyword(text string, keywords []string) bool {
	for _, kw := range keywords {
		if strings.Contains(text, strings.ToLower(kw)) {
			return true
		}
	}
	return false
}

// SaveArticles persists articles, returning the ids of newly inserted rows.
func (s *Scraper) SaveArticles(articles []*db.Article) []int64 {
	var ids []int64
	for _, a := range articles {
		id, err := s.db.SaveArticle(a)
		if err == nil && id != 0 {
			ids = append(ids, id)
		}
	}
	return ids
}
